File: renderers/android/src/main/kotlin/dev/varn/gui/VarnKeys.kt
package dev.varn.gui

import android.view.KeyEvent
import org.json.JSONObject

/**
 * The name each key is reported under, which is the one published set of names for them.
 *
 * A browser names a key in `KeyboardEvent.key`, iOS gives a HID usage and Android a key code, and the
 * three agree about nothing. The browser's names are what the tree reads, so the other two are mapped
 * onto them rather than each reporting whatever its own platform calls a key.
 */
object VarnKeys {
    private val NAMED = mapOf(
        KeyEvent.KEYCODE_ENTER to "Enter",
        KeyEvent.KEYCODE_NUMPAD_ENTER to "Enter",
        KeyEvent.KEYCODE_ESCAPE to "Escape",
        KeyEvent.KEYCODE_TAB to "Tab",
        KeyEvent.KEYCODE_DEL to "Backspace",
        KeyEvent.KEYCODE_FORWARD_DEL to "Delete",
        KeyEvent.KEYCODE_SPACE to " ",
        KeyEvent.KEYCODE_DPAD_UP to "ArrowUp",
        KeyEvent.KEYCODE_DPAD_DOWN to "ArrowDown",
        KeyEvent.KEYCODE_DPAD_LEFT to "ArrowLeft",
        KeyEvent.KEYCODE_DPAD_RIGHT to "ArrowRight",
        KeyEvent.KEYCODE_MOVE_HOME to "Home",
        KeyEvent.KEYCODE_MOVE_END to "End",
        KeyEvent.KEYCODE_PAGE_UP to "PageUp",
        KeyEvent.KEYCODE_PAGE_DOWN to "PageDown",
    )

    /** Answers what one key is called and what was held down with it. */
    fun payload(code: Int, key: KeyEvent): JSONObject {
        val named = NAMED[code] ?: written(key)

        return JSONObject()
            .put("key", named)
            .put("shift", key.isShiftPressed)
            .put("ctrl", key.isCtrlPressed)
            .put("alt", key.isAltPressed)
            .put("meta", key.isMetaPressed)
            .put("repeat", key.repeatCount > 0)
    }

    /** Answers the character a key stands for, or that nothing here knows what it is. */
    private fun written(key: KeyEvent): String {
        val character = key.getUnicodeChar(key.metaState)

        if (character == 0) {
            return "Unidentified"
        }

        return character.toChar().toString()
    }
}

File: renderers/ios/VarnKeys.swift
import UIKit

/// A view that reports a key it was given, which is what a key event reaches on this platform.
///
/// UIKit delivers a key press up the responder chain from whatever has focus, so a view is told about
/// one only while it is the first responder. That is the same contract the other two platforms have and
/// the one the tree is written to: a key reaches whatever has focus.
protocol VarnKeyed: UIResponder {
    var onKey: ((String, [String: Any]) -> Void)? { get set }
}

/// A view that says when the keyboard has reached it and when it has gone.
///
/// A control the platform draws paints its own focus ring. One the engine draws cannot, since nothing
/// told it, and using a keyboard is then a reader pressing keys at a screen that never says where they
/// are. `wanted` is what makes a box a stop on the way through at all.
protocol VarnFocusing: UIView {
    var onFocusChange: ((Bool) -> Void)? { get set }
    var wanted: Bool { get set }
}

/// The name each key is reported under, which is the one published set of names for them.
///
/// A browser names a key in `KeyboardEvent.key`, iOS gives a HID usage and Android a key code, and the
/// three agree about nothing. The browser's names are what the tree reads, so the other two are mapped
/// onto them rather than each reporting whatever its own platform calls a key.
enum VarnKeys {
    private static let named: [UIKeyboardHIDUsage: String] = [
        .keyboardReturnOrEnter: "Enter",
        .keypadEnter: "Enter",
        .keyboardEscape: "Escape",
        .keyboardTab: "Tab",
        .keyboardDeleteOrBackspace: "Backspace",
        .keyboardDeleteForward: "Delete",
        .keyboardSpacebar: " ",
        .keyboardUpArrow: "ArrowUp",
        .keyboardDownArrow: "ArrowDown",
        .keyboardLeftArrow: "ArrowLeft",
        .keyboardRightArrow: "ArrowRight",
        .keyboardHome: "Home",
        .keyboardEnd: "End",
        .keyboardPageUp: "PageUp",
        .keyboardPageDown: "PageDown",
    ]

    /// Answers what one key is called and what was held down with it.
    static func payload(of key: UIKey) -> [String: Any] {
        let flags = key.modifierFlags

        return [
            "key": named[key.keyCode] ?? (key.charactersIgnoringModifiers.isEmpty
                ? "Unidentified" : key.charactersIgnoringModifiers),
            "shift": flags.contains(.shift),
            "ctrl": flags.contains(.control),
            "alt": flags.contains(.alternate),
            "meta": flags.contains(.command),
            "repeat": false,
        ]
    }

    /// Reports every key of a press, answering whether anything was listening for one.
    static func report(_ presses: Set<UIPress>, as name: String, to view: VarnKeyed) -> Bool {
        guard let onKey = view.onKey else {
            return false
        }

        var told = false

        for press in presses {
            guard let key = press.key else {
                continue
            }

            onKey(name, payload(of: key))
            told = true
        }

        return told
    }
}
